Core/Src/main.c

```diff
@@ -134,16 +142,45 @@ void SystemClock_Config(void)
   {
     Error_Handler();
   }
-  PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_USART2;
-  PeriphClkInit.Usart2ClockSelection = RCC_USART2CLKSOURCE_PCLK1;
-  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK)
-  {
-    Error_Handler();
-  }
 }
 
 /* USER CODE BEGIN 4 */
+uint16_t readADC (uint8_t channel){
+
+
+	uint8_t receivedData[2] = {0};
+
+    uint8_t start_bit = 0x01;
+    uint8_t sglDiff = 1;      // singleended
+    uint8_t D2 = (channel >> 2) & 0x01;  // extract bit 2 of channel
+    uint8_t D1 = (channel >> 1) & 0x01;  // bit 1
+    uint8_t D0 = channel & 0x01;
+
+    uint8_t transmitData[3];
+    transmitData[0] = start_bit; //start
+    transmitData[1] = (sglDiff << 7) | (D2 << 6) | (D1 << 5) | (D0 << 4); //combine
+    transmitData[2] = 0x00;      //dummy byte
 
+	HAL_GPIO_WritePin(CS_PORT, CS_PIN, GPIO_PIN_RESET); //start transmission going low
+	HAL_SPI_TransmitReceive(&hspi1, transmitData, receivedData, 3, HAL_MAX_DELAY);
+	HAL_GPIO_WritePin(CS_PORT, CS_PIN, GPIO_PIN_SET);
+
+	uint16_t result = ((receivedData[1] & 0x0F) << 8) | receivedData[2];
+    return result;
+
+}
+
+void setPWMDutyCycle(uint16_t adc_value) {
+	uint32_t period = __HAL_TIM_GET_AUTORELOAD(&htim1);
+	uint32_t min_duty_cycle = 0.05 * period;
+	uint32_t max_duty_cycle = 0.10 * period;
+
+
+	uint32_t compare_value = min_duty_cycle + ((adc_value * (max_duty_cycle - min_duty_cycle)) / 1023);
+
+
+	__HAL_TIM_SET_COMPARE(&htim1, TIM_CHANNEL_1, compare_value);
+}
```

Comment:
Great Work! You figure out how to correctly configure the PWM frequency, all the calculations are correct! I have a couple of suggestions that could make your code more robust, but you don't have to change the code.
1. pull the CS line high at the beginning of the program.
2. add an input boundary check in the `readADC` function to make sure the input is valid.